views/js/runner/proxy/cache/proxy.js

```diff
@@ -133,6 +128,178 @@ define([
+                        testContext = testNavigator.navigate(
+                                actionParams.direction,
+                                actionParams.scope,
+                                actionParams.ref
+                            );
+
+                        //we are really not able to navigate
+                        if(!testContext || !testContext.itemIdentifier || !self.hasItem(testContext.itemIdentifier)){
+                            offlineNavError = new Error(__('Unable to select the next item due to connectivity issues'));
+                            _.assign(offlineNavError, {
+                                success : false,
+                                source: 'navigator',
+                                purpose: 'proxy',
+                                type: 'Item not found',
+                                code : 404
+                            });
+                            throw offlineNavError;
+                        }
+                        return {
+                            success : true,
+                            testContext : testContext
+                        };
+                    }
+
+                    //at worst, we have saved the action and just want to continue
+                    return {
+                        success: true
+                    };
+                });
+            };
+
+            /**
+             * Request/Offlin strategy :
+             *
+             * ├─ Online
+             * │  └─ run the request
+             * │    ├─ request ok
+             * │    └─ request fails
+             * │       └─ run the offline action
+             * └── Offline
+             *    └─ send a telemetry request (connection could be back)
+             *      ├─ request ok
+             *      │  └─ sync data
+             *      │     └─  run the request (back to the tree root)
+             *      └─ request fails
+             *         └─ run the offline action
+             *
+             * @param {String} action - the action name (ie. move, skip, timeout)
+             * @param {Object} actionParams - the parameters sent along the action
+             * @returns {Promise} resolves with the action result
+             */
+            this.requestNetworThenOffline = function requestNetworThenOffline(url, action, actionParams){
+
+                //perform the request, but fallback on offline if the request itself fails
+                var runRequetThenOffline = function runRequetThenOffline(){
+                    return self.request(url, actionParams).then(function(result){
+                        //if the request fails, we should be offline
+                        if(self.isOffline()){
+                            return self.offlineAction(action, actionParams);
+                        }
+                        return result;
+                    });
+                };
```

Comment:
The promise rejection is not catched, so the connectivity error should only appear on resolve?